wallet-new/server/Cardano/Wallet/API/V1/LegacyHandlers/Addresses.hs

```diff
@@ -51,10 +56,40 @@ newAddress NewAddress {..} = do
               >>= migrate
 
 -- | Verifies that an address is base58 decodable.
-verifyAddress :: Monad m => Text -> m (WalletResponse AddressValidity)
-verifyAddress address =
-    case decodeTextAddress address of
-        Right _ ->
-            return $ single $ AddressValidity True
-        Left _  ->
-            return $ single $ AddressValidity False
+getAddress
+    :: (MonadThrow m , V0.MonadWalletLogic ctx m)
+    => Text
+    -> m (WalletResponse AddressInfo)
+getAddress addrText = do
+    addr <- either
+        (throwM . InvalidAddressFormat)
+        pure
+        (decodeTextAddress addrText)
+
+    ss <- V0.askWalletSnapshot
+
+    let mOurAddr = ss ^? V0.wsBalances . at addr
```

Comment:
Unfortunately, this might not be enough. If you look at the comment inside `Storage.hs`, it says:

```
      -- | Current balances of wallet's addresses
      -- NB: @_wsBalances@ depends on @_wsUtxo@,
      -- it's forbidden to update @_wsBalances@ without @_wsUtxo@.
    , _wsBalances        :: !WalletBalances
```

This means that it will include only the addresses which has actual money associated to them, and this info is kept updated with the UTXO, which means that `wsBalances` won't contain an address `A` if we spent all the available balance for it. But as part of this endpoint, we probably want to return it straight-away.

As @edsko 's new `kernel` will eventually force us to rewrite all these handlers, I suggest we go with the super dumb solution, and reuse the "lego blocks" provided as part of `Storage.hs`. Namely, we could use `getWalletAddresses` to get _all the wallet ids_ known to the backend, and then iterate over those and call `getWAddresses`. If you look at the signature:

```haskell
getWAddresses :: AddressLookupMode
              -> CId Wal
              -> Query [AddressInfo]
getWAddresses mode wid =
```

You can pass an `AddressLookupMode`. If you pass `Ever` this should give you back a `[AddressInfo]` even if this address was deleted (I have to admit the notion of a deleted address is a bit of a grey area to me, but a quick detour in the codebase should clarify). I think point we _might_ be in a good shape to answer the original query, otherwise we might need to do more db access.

It's a bit convoluted, yes (so is the old data layer) but at least we don't invent any new logic, which means less porting work later 😉

Reply:
Ahh, thanks!

I found it was quite difficult to navigate the data relationships :frowning_face: Tfw no SQL tables to parse relationships out of...